win32ss/gdi/ntgdi/freetype.c

```diff
@@ -35,6 +35,186 @@
+    PDC dc;
+} FONTLINK_CHAIN, *PFONTLINK_CHAIN;
+
+typedef struct _FONTLINK_CACHE
+{
+    LIST_ENTRY ListEntry;
+    LOGFONTW LogFont;
+    FONTLINK_CHAIN Chain;
+} FONTLINK_CACHE, *PFONTLINK_CACHE;
+
+#define FONTLINK_DEFAULT_CHAR 0x30FB // U+30FB (KATAKANA MIDDLE DOT)
+
+static DWORD s_nFontLinkControl = 0; // FIXME: What to do?
+static DWORD s_chFontLinkDefaultChar = FONTLINK_DEFAULT_CHAR;
+
+#define MAX_FONTLINK_CACHE 128
+static RTL_STATIC_LIST_HEAD(g_FontLinkCache); // The list of FONTLINK_CACHE
+static LONG g_nFontLinkCacheCount = 0;
+
+static SIZE_T
+SZZ_GetSize(_In_ PCZZWSTR pszz)
+{
+    SIZE_T ret = 0, cch;
+    const WCHAR *pch = pszz;
+    while (*pch)
+    {
+        cch = wcslen(pch) + 1;
+        ret += cch;
+        pch += cch;
+    }
+    ++ret;
+    return ret * sizeof(WCHAR);
+}
+
+static NTSTATUS
+FontLink_LoadSettings(VOID)
+{
+    NTSTATUS Status;
+    HKEY hKey;
+    DWORD cbData;
+
+    s_nFontLinkControl = 0;
+    s_chFontLinkDefaultChar = FONTLINK_DEFAULT_CHAR;
+
+    Status = RegOpenKey(
+        L"\\Registry\\Machine\\Software\\Microsoft\\Windows NT\\CurrentVersion\\FontLink",
+        &hKey);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    cbData = sizeof(s_nFontLinkControl);
+    RegQueryValue(hKey, L"FontLinkControl", REG_DWORD, &s_nFontLinkControl, &cbData);
+
+    cbData = sizeof(s_chFontLinkDefaultChar);
+    RegQueryValue(hKey, L"FontLinkDefaultChar", REG_DWORD, &s_chFontLinkDefaultChar, &cbData);
+
+    ZwClose(hKey); // Close the registry key
+    return STATUS_SUCCESS;
+}
+
+static VOID
+FontLink_Destroy(_Inout_ PFONTLINK pLink)
+{
+    ASSERT(pLink);
+    ExFreePoolWithTag(pLink, TAG_FONT);
+}
+
+static VOID
+FontLink_Chain_Free(
+    _Inout_ PFONTLINK_CHAIN pChain)
+{
+    PLIST_ENTRY Entry;
+    PFONTLINK pLink;
+
+    if (!pChain->szFaceName[0]) // Not populated yet
+        return;
+
+    ExFreePoolWithTag(pChain->pszzFontLink, TAG_FONT);
+
+    while (!IsListEmpty(&pChain->FontLinkList))
+    {
+        Entry = RemoveHeadList(&pChain->FontLinkList);
+        pLink = CONTAINING_RECORD(Entry, FONTLINK, ListEntry);
+        FontLink_Destroy(pLink);
+    }
+}
+
+static inline VOID
+FontLink_AddCache(
+    _In_ PFONTLINK_CACHE pCache)
+{
+    PLIST_ENTRY Entry, Head = &g_FontLinkCache;
+    ++g_nFontLinkCacheCount;
+    InsertHeadList(Head, &pCache->ListEntry);
+
+    if (g_nFontLinkCacheCount > MAX_FONTLINK_CACHE)
+    {
+        ASSERT(!IsListEmpty(Head));
+        Entry = RemoveTailList(Head);
+        --g_nFontLinkCacheCount;
+        pCache = CONTAINING_RECORD(Entry, FONTLINK_CACHE, ListEntry);
+        FontLink_Chain_Free(&pCache->Chain);
+        ExFreePoolWithTag(pCache, TAG_FONT);
+    }
+}
+
+static VOID
+FontLink_Chain_Finish(
+    _Inout_ PFONTLINK_CHAIN pChain)
+{
+    PFONTLINK_CACHE pCache;
+    PLIST_ENTRY Entry;
+
+    if (!pChain->szFaceName[0]) // Not populated yet
+        return;
+
+    pCache = ExAllocatePoolWithTag(PagedPool, sizeof(FONTLINK_CACHE), TAG_FONT);
```

Comment:
```suggestion
    pCache = ExAllocatePoolZero(PagedPool, sizeof(FONTLINK_CACHE), TAG_FONT);
```
unless you want to keep the contents of the cache undefined and hope for the rest of the function (and the AddCache) helper to correctly initialize them :eyes:

Reply:
No zero-filling required.